### ISSUE TITLE
Wayland connection broke. Did the Wayland compositor die?

### ISSUE THREAD

**Issue by mitinarseny** [ifreund/river#332]
I faced a bug, when [telegram-desktop](https://github.com/telegramdesktop/tdesktop) (it is built on gtk+3) crashes, saying:
```sh
$ telegram-desktop
// other logs ...
Invalid return value 0 for stream protocol
Invalid return value 0 for stream protocol
Invalid return value 0 for stream protocol
Invalid return value 0 for stream protocol
The Wayland connection broke. Did the Wayland compositor die?
```

Also, `wl-copy` from [wl-clipboard](https://github.com/bugaevc/wl-clipboard) hangs if run in foreground:
```sh
$ echo abc | wl-copy -f
# ... it never ends
```

Could it be a problem with river?

**Comment by ifreund** [comment]
> I faced a bug, when [telegram-desktop](https://github.com/telegramdesktop/tdesktop) (it is built on gtk+3) crashes, saying:

Pretty sure you mean qt5. Anyhow, I can't reproduce this crash, does it happen under e.g. sway as well?

> Also, `wl-copy` from [wl-clipboard](https://github.com/bugaevc/wl-clipboard) hangs if run in foreground:

I don't know what the intended behavior of wl-copy is here, but I'm pretty sure river isn't doing anything wrong. 

> Could it be a problem with river?

There's a simple way to test this, check if the behavior is different under some other compositor (e.g. sway).

Anyhow, there's no clear issue reported here so I'm going to close this.

**Comment by mitinarseny** [comment]
Isn't an unavailability of wayland connection a clear issue? 

In sway this bug didn't go away. But it does not mean, that everything is ok: the unavailability of connection is a problem of compositor, not the application.

Here is how I reporoduce this bug with telegram in case you are familiar with it:
* Open telegram-desktop
* Play any song in any of your chats
* Make it playing repetitive (so that the same song starts when it ends)
* Send telegram to a separate workspace
* Go to another workspace, do something unrelated for ~10 mins
* Switch back to workspace with telegram
* Boom! It has crashed

Here are the logs of 
```sh
$ WAYLAND_DEBUG=1 telegram-desktop
// ...
[930150.616] wl_buffer@22.release()
[930159.008]  -> wl_surface@19.frame(new id wl_callback@69957)
[930159.022]  -> wl_surface@19.attach(wl_buffer@22, 0, 0)
[930159.033]  -> wl_surface@19.damage(731, 51, 1185, 6)
[930159.044]  -> wl_surface@19.commit()
[930159.195] wl_buffer@22.release()
[930167.500]  -> wl_surface@19.frame(new id wl_callback@69958)
[930167.516]  -> wl_surface@19.attach(wl_buffer@22, 0, 0)
[930167.524]  -> wl_surface@19.damage(731, 51, 1185, 6)
[930167.531]  -> wl_surface@19.commit()
[930167.731] wl_buffer@22.release()
[930176.207]  -> wl_surface@19.frame(new id wl_callback@69959)
[930176.231]  -> wl_surface@19.attach(wl_buffer@22, 0, 0)
[930176.242]  -> wl_surface@19.damage(731, 51, 1185, 6)
[930176.252]  -> wl_surface@19.commit()
[930176.432] wl_buffer@22.release()
// at this point logs start to be VERY frequent
// all of following had happened in less than 1 sec
[961828.516] wl_keyboard@9.enter(2461, wl_surface@19, array)
[961828.539] wl_keyboard@9.modifiers(2462, 64, 0, 0, 0)
[961828.565] wl_data_device@13.data_offer(new id wl_data_offer@1815825600)
[961828.576] wl_data_offer@4278190081.offer("text/plain;charset=utf-8")
[961828.585] wl_data_offer@4278190081.offer("UTF8_STRING")
[961828.591] wl_data_device@13.selection(wl_data_offer@4278190081)
[961828.600] zwp_primary_selection_device_v1@14.data_offer(new id zwp_primary_selection_offer_v1@1815825472)
[961828.609] zwp_primary_selection_offer_v1@4278190080.offer("text/plain;charset=utf-8")
[961828.616] zwp_primary_selection_offer_v1@4278190080.offer("UTF8_STRING")
[961828.621] zwp_primary_selection_device_v1@14.selection(zwp_primary_selection_offer_v1@4278190080)
[961828.627] wl_pointer@3.enter(2464, wl_surface@19, 1077.347656, 416.054688)
[961828.652]  -> wl_pointer@3.set_cursor(2464, wl_surface@39, 1, 1)
[961828.667]  -> wl_surface@39.set_buffer_scale(1)
[961828.673]  -> wl_surface@39.attach(wl_buffer@40, 0, 0)
[961828.685]  -> wl_surface@39.damage(0, 0, 10, 16)
[961828.696]  -> wl_surface@39.commit()
[961828.701] wl_pointer@3.frame()
[961828.714] xdg_toplevel@26.configure(1916, 1029, array)
[961828.726] xdg_surface@25.configure(2460)
[961828.772]  -> wl_display@1.sync(new id wl_callback@69960)
[961828.792]  -> xdg_toplevel@26.set_min_size(452, 501)
[961828.801]  -> xdg_toplevel@26.set_max_size(0, 0)
[961828.810]  -> xdg_surface@25.ack_configure(2460)
[961829.151] wl_surface@39.enter(wl_output@16)
[961829.166]  -> wl_surface@39.set_buffer_scale(1)
[961829.174]  -> wl_surface@39.attach(wl_buffer@40, 0, 0)
[961829.186]  -> wl_surface@39.damage(0, 0, 10, 16)
[961829.200]  -> wl_surface@39.commit()
[961829.208] wl_display@1.delete_id(69960)
[961829.220] wl_callback@69960.done(2464)
[961838.014] wl_display@1.delete_id(69959)
[961838.036] wl_display@1.delete_id(69958)
// many times:
// [...] wl_display@1.delete_id(...)
[961849.122] wl_display@1.delete_id(67916)
[961849.127] wl_display@1.delete_id(67915)
[961849.132] wl_display@1.delete_id(67914)
[961851.638]  -> wl_surface@19.frame(new id wl_callback@69959)
[961851.669]  -> wl_surface@19.attach(wl_buffer@22, 0, 0)
[961851.683]  -> wl_surface@19.damage(0, 0, 1916, 1029)
[961851.695]  -> wl_surface@19.commit()
[961851.756] wl_display@1.delete_id(67913)
[961851.763] wl_display@1.delete_id(67912)
[961851.768] wl_display@1.delete_id(67911)
// many times:
// [...] wl_display@1.delete_id(...)
[961866.168] wl_display@1.delete_id(65358)
[961866.174] wl_display@1.delete_id(65357)
[961867.797]  -> wl_surface@19.frame(new id wl_callback@65357)
[961867.816]  -> wl_surface@19.attach(wl_buffer@32, 0, 0)
[961867.827]  -> wl_surface@19.damage(731, 51, 1185, 6)
[961867.841]  -> wl_surface@19.damage(1904, 113, 12, 867)
[961867.854]  -> wl_surface@19.damage(777, 992, 1007, 28)
[961867.866]  -> wl_surface@19.commit()
[961867.923] wl_display@1.delete_id(65356)
[961867.931] wl_display@1.delete_id(65355)
[961867.936] wl_display@1.delete_id(65354)
// many times:
// [...] wl_display@1.delete_id(...)
[961883.616] wl_display@1.delete_id(62290)
[961883.621] wl_display@1.delete_id(62289)
[961883.626] wl_display@1.delete_id(62288)
[961883.725]  -> wl_shm@4.create_pool(new id wl_shm_pool@62288, fd 22, 7886256)
[961883.739]  -> wl_shm_pool@62288.create_buffer(new id wl_buffer@62289, 0, 1916, 1029, 7664, 0)
[961888.254]  -> wl_surface@19.frame(new id wl_callback@62290)
[961888.274]  -> wl_surface@19.attach(wl_buffer@62289, 0, 0)
[961888.286]  -> wl_surface@19.damage(0, 21, 72, 30)
[961888.299]  -> wl_surface@19.damage(0, 51, 72, 6)
[961888.311]  -> wl_surface@19.damage(731, 51, 1185, 6)
[961888.322]  -> wl_surface@19.damage(0, 57, 72, 18)
[961888.332]  -> wl_surface@19.damage(1904, 113, 12, 867)
[961888.342]  -> wl_surface@19.damage(777, 992, 1007, 28)
[961888.354]  -> wl_surface@19.commit()
[961888.407] wl_display@1.delete_id(62287)
[961888.414] wl_display@1.delete_id(62286)
[961888.419] wl_display@1.delete_id(62285)
// many times:
// [...] wl_display@1.delete_id(...)
[961894.355] wl_display@1.delete_id(61096)
[961894.360] wl_display@1.delete_id(61095)
[961894.365] wl_display@1.delete_id(61094)
The Wayland connection broke. Did the Wayland compositor die?
```

**Comment by ifreund** [comment]
> Isn't an unavailability of wayland connection a clear issue?

Sorry, I should have been more clear and said "there's no clear issue *with river* reported".

> In sway this bug didn't go away. But it does not mean, that everything is ok: the unavailability of connection is a problem of compositor, not the application.

Not necessarily, observe: 

This snippet of the log tells me that `telegram-desktop` is not using frame callbacks properly:
```
[930159.008]  -> wl_surface@19.frame(new id wl_callback@69957)
[930159.022]  -> wl_surface@19.attach(wl_buffer@22, 0, 0)
[930159.033]  -> wl_surface@19.damage(731, 51, 1185, 6)
[930159.044]  -> wl_surface@19.commit()
[930159.195] wl_buffer@22.release()
[930167.500]  -> wl_surface@19.frame(new id wl_callback@69958)
[930167.516]  -> wl_surface@19.attach(wl_buffer@22, 0, 0)
[930167.524]  -> wl_surface@19.damage(731, 51, 1185, 6)
[930167.531]  -> wl_surface@19.commit()
[930167.731] wl_buffer@22.release()
[930176.207]  -> wl_surface@19.frame(new id wl_callback@69959)
[930176.231]  -> wl_surface@19.attach(wl_buffer@22, 0, 0)
[930176.242]  -> wl_surface@19.damage(731, 51, 1185, 6)
[930176.252]  -> wl_surface@19.commit()
[930176.432] wl_buffer@22.release()
```
It asks for a frame callback and then, instead of waiting for the `done` event on the callback to be received, it attaches and commits a buffer anyways. This in an of itself is not strictly speaking a problem, the bug here is that the client then requests another frame callback again and again despite no `done` event on any frame callback having been received. Eventually, this client behavior will cause bad things to happen, such as running out of IDs or memory.

**Comment by ifreund** [comment]
```
17:23 <ifreund> great: https://github.com/ifreund/river/issues/332#issuecomment-864570512
17:24 <ifreund> I hope that's specifc to telegram-desktop, not all QT-based things
...
17:35 <alebastr> ifreund: that's affecting all qt-based things
17:37 <ifreund> *sigh*
17:37 <alebastr> good thing is that it's *almost* fixed in qt 6.2. there's a couple more pending patches to address very special cases
17:41 <ifreund> well at least things are heading in the right direction...
```
